model : support LiquidAI LFM2 hybrid family

tdakhran opened the pull request:
### Summary

Add support for [LiquidAI LFM2](https://huggingface.co/collections/LiquidAI/lfm2-686d721927015b2ad73eaa38) model family. For more information about models, please read [the blog post](https://www.liquid.ai/blog/liquid-foundation-models-v2-our-second-series-of-generative-ai-models).

- Support hybrid LFM2-350M, LFM2-700M, and LFM2-1.2B models.
- Add `ShortConvBlock`.
- Modify `construct_transformer` to construct hybrid architectures.
- Move FeedForward to avoid cyclid dependency

Instructions are in `examples/models/lfm2/README.md`.

### Test plan
All commands in `README.md` are tests.
```
❯ python -m examples.models.llama.runner.native \
  --model lfm2_700m \
  --pte lfm2_700m_8da4w.pte \
  --tokenizer ~/.cache/huggingface/hub/models--LiquidAI--LFM2-700M/snapshots/ab260293733f05dd4ce22399bea1cae2cf9b272d/tokenizer.json \
  --tokenizer_config ~/.cache/huggingface/hub/models--LiquidAI--LFM2-700M/snapshots/ab260293733f05dd4ce22399bea1cae2cf9b272d/tokenizer_config.json \
  --prompt "<|startoftext|><|im_start|>user\nWho are you?<|im_end|>\n<|im_start|>assistant\n" \
  --params examples/models/lfm2/config/lfm2_700m_config.json \
  --max_len 128 \
  -kv \
  --temperature 0.3

...
I'm an AI designed to assist with generating text based on the prompts you provide. I'm a type of language model, but I don't have a physical form or consciousness. I operate based on complex algorithms and vast amounts of training data. How can I help you today? If you have a specific question or need assistance with something, feel free to ask!
...
```

tdakhran committed:
model : support LiquidAI LFM2 hybrid family

Add support for [LiquidAI LFM2](https://huggingface.co/collections/LiquidAI/lfm2-686d721927015b2ad73eaa38) model family.
For more information about models, please read [the blog post](https://www.liquid.ai/blog/liquid-foundation-models-v2-our-second-series-of-generative-ai-models).

- Support hybrid LFM2-350M, LFM2-700M, and LFM2-1.2B models.
- Add `ShortConvBlock`.
- Modify `construct_transformer` to construct hybrid architectures.
- Move FeedForward to avoid cyclid dependency

jackzhxng reviewed (approved):
Looks good to me, can you try running with the C++ runner? https://github.com/pytorch/executorch/blob/main/examples/models/llama/README.md#step-3-run-on-your-computer-to-validate

mergennachin commented:
Please rebase to main and resolve the conflicts

Don't merge yet, I will have to import internally and run a few internal tests.

tdakhran committed:
Address PR feedback

tdakhran commented:
@jackzhxng @mergennachin  thanks for the review, I've rebased and addressed the feedback.

All commands in README.md work, including the C++ runner. However, the C++ runner output differs from the Python runner.

jackzhxng reviewed (approved):
Thanks the work @tdakhran!

kimishpatel commented on examples/models/llama/llama_transformer.py:
```diff
@@ -255,6 +244,13 @@ def construct_transformer(model_args: ModelArgs) -> Transformer:
     cls = ATTENTION_REGISTRY[model_args.attention_type]
     for layer_id in range(model_args.n_layers):
+        # hybrid models define layer_types
+        if model_args.layer_types and model_args.layer_types[layer_id] == "conv":
+            layers.append(
+                ShortConvBlock(dim=model_args.dim, hidden_dim=model_args.hidden_dim, norm_eps=model_args.norm_eps)
+            )
+            continue
```
nit: can you not do if... else? why continue pattern?

kimishpatel commented:
left some small comments. is there any perf numbers on ios or android?

tdakhran committed:
Address PR feedback

mergennachin commented:
@tdakhran - still a few minor lints -- https://github.com/pytorch/executorch/actions/runs/17426424686/job/49500257182?pr=13805

tdakhran committed:
More linting

tdakhran commented:
> @tdakhran - still a few minor lints -- https://github.com/pytorch/executorch/actions/runs/17426424686/job/49500257182?pr=13805

Had an impression that `lintrunner -a` fixes them all, pushed the fix. Thanks for noticing!

Commit:
Merge branch 'main' into tarek/feat/lfm2_upstream

jackzhxng committed:
Merge branch 'main' into tarek/feat/lfm2_upstream

jackzhxng committed:
Merge branch 'main' into tarek/feat/lfm2_upstream

kimishpatel commented on examples/models/lfm2/short_conv.py:
```diff
@@ -0,0 +1,110 @@
+        #  So, assuming prefill is done on an empty cache, concatenating conv_state to the beginning of the sequence acts similary to
+        ## using nn.Conv1d(padding=L_cache-1) (for prefill) without no manual padding.
+        ## However, the manual padding has the added benefit of being correct during decode, when the cache is not initialized to 0.
+        Bx = torch.cat(
+            [self.conv_state, Bx], dim=-1
+        )  # (batch_size, dim, seq_len + L_cache - 1)
+
+        ## Update the conv_state
+        new_conv_state = Bx[
+            ..., -(self.L_cache - 1) :
+        ]  # (batch_size, dim, L_cache - 1)
+        with torch.no_grad():
+            self.conv_state.copy_(new_conv_state)
```
isnt this 
```
        new_conv_state = Bx[
            ..., -(self.L_cache - 1) :
        ]  # (batch_size, dim, L_cache - 1)
```
effectively slicing Bx for last `L_cache` entries after concat, which sounds like sliding window of size `L_cache`?
